feat: TAP Agent

aasseman opened the pull request:
Sorry, it's a big one. Perhaps we could set up a call to discuss this PR.

The TAP Agent is a new daemon that indexers will have to run alongside the indexer-service instances. It:
- Monitors receipts being saved by the indexer-service instances into the DB (through pg_notify).
- Keeps track of the unaggregated fees for each Account, where an Account is (indexer, sender, allocation).
- Handles the receipt validation checks (because the indexer-service is designed to do barely any checks).
- Handles RAV (Receipt Aggregate Vouchers) requests whenever the total unaggregated fees surpass a trigger amount.
- Automatically does one last RAV request when an Account relationship ends (Allocation closed, for example).

What it should, but *does not do yet*:
- Check the receipt values against the Agora system
- Signal the indexer-service that a sender should be banned (not enough in Escrow, too many invalid receipts, etc)
- A lot of other, smaller, but important things, that are marked as `TODO` in the code...

I'm not particularly proud of the overall design, and it also shone some light onto design shortcomings in `tap_core` that we did not foresee while designing it "in a vacuum". That's why there is some weirdness in the TAP adapters implementation such as https://github.com/graphops/indexer-service-rs/blob/20e48b1191a8d39f02e32200ef75fc065470cdd6/tap_agent/src/tap/receipt_checks_adapter.rs#L64-L76.

My goal here is to get something through the door so that we can continue improving this incrementally together. I propose that we try moving as many review issues into actual issues that we tackle later.

Jannis commented on common/src/tap_manager.rs:
```diff
@@ -86,14 +86,20 @@ impl TapManager {
+                INSERT INTO scalar_tap_receipts (allocation_id, sender_address, timestamp_ns, value, receipt)
+                VALUES ($1, $2, $3, $4, $5)
             "#,
             format!("{:?}", allocation_id)
                 .strip_prefix("0x")
                 .unwrap()
                 .to_owned(),
+                receipt_signer
+                .to_string()
+                .strip_prefix("0x")
+                .unwrap()
+                .to_owned(),
```
In both these cases it would be better to use [`trim_start_matches("0x")`](https://doc.rust-lang.org/std/string/struct.String.html#method.trim_start_matches). `strip_prefix` returns `None` if the allocation ID is lacking a `0x` prefix. If we're doing validations anywhere, we should do it in a way that reports an error back to the client and does not crash the service.

aasseman replied:
Ah, seems I didn't RTFM hard enough! Yes that'd be much better indeed, thanks!

aasseman replied:
[f7447f7](https://github.com/graphops/indexer-service-rs/pull/83/commits/f7447f7f27a8436b74b4d20ce7abb878e576c354)

hopeyen commented on tap_agent/src/config.rs:
```diff
@@ -0,0 +1,320 @@
+    )]
+    pub rav_request_timestamp_buffer_ns: u64,
+
+    // TODO: Remove this whenever the the gateway registry is ready
```
how do we expect indexers to have this file? does senders register or host a discoverable aggregator endpoint somewhere? would you please provide an example file I could run the service with?

aasseman replied:
It will be in the form of a smart contract. That's the current plan/suggestion I got from @Theodus . But no work started on this yet I believe.

aasseman replied:
PS: the file is just to be able to get going for testing, basically.

hopeyen commented on tap_agent/src/config.rs:
```diff
@@ -0,0 +1,320 @@
+    as Receipt Aggregate Voucher (RAV) requests for an indexer."
+)]
+#[command(author, version, about, long_about = None, arg_required_else_help = true)]
+pub struct Cli {
```
Some of these are duplicates of `indexer-service CLI`, would it make sense to refactor the duplicates to `indexer-common`?

aasseman replied:
Yup! Not sure there is an elegant way to do that though.

aasseman replied:
Actually in k8s, the preferred way of passing secrets is using env vars. Unless you are willing to integrate your software directly with the k8s or your cloud's secrets API.
My favorite config implementation is that of `graph-node`. It uses TOML, but also does something like `envsubst` over the file before loading it.

aasseman replied:
> https://www.cloudtruth.com/blog/the-pitfalls-of-using-environment-variables-for-config-and-secrets suggests that environment variables for secrets is insecure.

I'm not sure this really applies to containers though. This is still one of the recommended ways in k8s for ex: https://kubernetes.io/docs/tasks/inject-data-application/distribute-credentials-secure/

hopeyen reviewed (commented):
Great work! still need to digest the `tap` details, would you mind sharing the flow diagram somewhere? 🙏

aasseman commented on migrations/20230912220523_tap_receipts.up.sql:
Yes, it's there, I just didn't change the version from main: https://github.com/graphops/indexer-service-rs/blob/407c17d3ecf766c8e37a2b7e4b9b1132136c4a33/migrations/20230912220523_tap_receipts.down.sql

Also, I modified the migrations in place instead of creating new ones. I thought at this stage of development we don't need to care about migrating the database. (tell me if that's the wrong thinking though)

aasseman commented on tap_agent/src/agent.rs:
```diff
@@ -0,0 +1,91 @@
+        SubgraphClient::new(
+            config
+                .network_subgraph
+                .network_subgraph_deployment
```
It's the new API introduced in #81

aasseman commented:
> Great work! still need to digest the `tap` details, would you mind sharing the flow diagram somewhere? 🙏

Working on it!

aasseman commented:
Sorry for the delay. Here's a simple diagram. Which part would you like more details for?

![tap_agent](https://github.com/graphops/indexer-service-rs/assets/33075224/c6ecb452-19b1-4644-84a6-0a6d5ce39e93)

Jannis commented on tap_agent/src/tap/account.rs:
```diff
@@ -0,0 +1,846 @@
+    }
+
+    async fn rav_requester_try(inner: &Arc<Inner>) -> anyhow::Result<()> {
+        loop {
```
Related to this: do we make sure that we don't request multiple RAVs for the same receipts in parallel if a new receipt comes in before the RAV comes back and we still have unaggregated fees > threshold?

aasseman replied:
Yes because we always (unless I missed it somewhere) check `Self::rav_requester_task_is_running` before starting the `rav_requester` in an async thread.

aasseman replied:
- We do make sure that we don't do RAV requests in parallel for any given (sender,allocation), by checking `Self::rav_requester_task_is_running` before spawning a RAV request task. Could be made more elegant I think.
- Each RAV request must contain the last RAV (unless it is the first RAV request). We basically aggregate the receipts into the latest RAV we got, and we get a new RAV out of that. That's why it is very important the the whole RAV request dance is serial.
- The TapManager will only aggregate receipts that have a timestamp more recent than the last RAV (though note that we are OK with multiple receipts having the same timestamp, they also have a nonce). To make sure that there is no confusion with the new receipts coming in *while* we're requesting a RAV, the TapManager will only aggregate receipts up to timestamps that are more than `timestamp_buffer` old: https://github.com/semiotic-ai/timeline-aggregation-protocol/blob/656fc15700f6d29299d78f9c790539cbddb68d02/tap_core/src/tap_manager/manager.rs#L179

  One issue I can see here is that we are not checking (in indexer-service) that new receipts have a reasonably accurate timestamp (let's say within 5 seconds of our own clock). So as it is right now it is not really safe :/.

aasseman replied:
Yup, created and issue: #92

Jannis reviewed (commented):
Made it 30% through. Will keep going but wanted to drop my comments so far. 😉

aasseman commented on migrations/20230915230734_tap_ravs.up.sql:
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE IF NOT EXISTS scalar_tap_latest_ravs (
```
Nah, not really. I picked a more explicit name (only stores the latest rav for any given (allocation, sender). But not necessary. I can shorten this.

aasseman replied:
Well, also I made a mistake since it doesn't match this: https://github.com/graphprotocol/indexer-rs/blob/f7447f7f27a8436b74b4d20ce7abb878e576c354/migrations/20230915230734_tap_ravs.down.sql#L1 ...

aasseman replied:
Fixed in 1ce92444b7287775bd4f2134bb08b06cdedce72b

aasseman commented on tap_agent/Cargo.toml:
```diff
@@ -0,0 +1,46 @@
+[package]
+name = "indexer_tap_agent"
+version = "0.1.0"
+edition = "2021"
+publish = false
+
+[[bin]]
+name = "indexer_tap_agent"
```
Will do!

aasseman replied:
Fixed in 604153493624338eeb700856c71b9735154b7bd0

aasseman commented on tap_agent/src/agent.rs:
```diff
@@ -0,0 +1,91 @@
+        Duration::from_secs(config.escrow_subgraph.escrow_syncing_interval),
+    );
+
+    // TODO: replace with a proper implementation once the gateway registry contract is ready
+    let sender_aggregator_endpoints = aggregator_endpoints::load_aggregator_endpoints(
+        config.tap.sender_aggregator_endpoints_file.clone(),
+    );
```
Sounds good!

aasseman replied:
Saved your comment in the appropriate issue: https://github.com/graphprotocol/indexer-rs/issues/84#issuecomment-1809245832

aasseman commented on tap_agent/src/main.rs:
```diff
@@ -0,0 +1,44 @@
+    let mut signal_sigint = signal(SignalKind::interrupt())?;
+    let mut signal_sigterm = signal(SignalKind::terminate())?;
+    tokio::select! {
+        _ = signal_sigint.recv() => debug !("Received SIGINT."),
+        _ = signal_sigterm.recv() => debug !("Received SIGTERM."),
```
:thinking: no.... Perhaps `cargo fmt` did that, though I think that's not what it is supposed to do in this situation....

aasseman replied:
Fixed in da11a93549335e555692ddd7624937d4b19d24ba

aasseman commented on tap_agent/src/tap/account.rs:
```diff
@@ -0,0 +1,846 @@
+}
+
+impl Account {
+    #[allow(clippy::too_many_arguments)]
```
:grimacing:

aasseman commented on tap_agent/src/tap/account.rs:
```diff
@@ -0,0 +1,846 @@
+    config: &'static config::Cli,
+}
+
+/// An Account is the relationship between the allocation and the sender in the context of a single
```
Yeah `SenderAllocationRelationship`  is probably better. I struggled with the naming here too :/

aasseman replied:
Fixed in 226a8ff3033261ce97dc755a267a8d646e3a4648

aasseman commented on tap_agent/src/main.rs:
```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use lazy_static::lazy_static;
```
One point for `once_cell` is that it could be added to the standard library: https://github.com/rust-lang/rfcs/pull/2788

aasseman commented on tap_agent/src/tap/account.rs:
```diff
@@ -0,0 +1,846 @@
+
+            let mut rav_requester_task = self.rav_requester_task.lock().await;
+            // TODO: consider making the trigger per sender, instead of per (sender, allocation).
+            if unaggregated_fees.value >= self.inner.config.tap.rav_request_trigger_value.into()
```
Yup, the grand total is kept in `tap_agent/src/tap/escrow_adapter.rs`.
The `TapManager` will subtract escrow there as it is running the receipt checks (for now that happens only in batch right before a RAV request). If it goes over the escrow the RAV request will fail.
We have yet to implement a way to relay the failure to the indexer services.

aasseman commented on tap_agent/src/tap/account.rs:
```diff
@@ -0,0 +1,846 @@
+            unaggregated_fees.last_id = new_receipt_notification.id;
+
+            let mut rav_requester_task = self.rav_requester_task.lock().await;
+            // TODO: consider making the trigger per sender, instead of per (sender, allocation).
```
Yes. Makes the code slightly more complex because the RAVs still have to be per (sender, allocation).

Would it make sense to have a strategy for which allocation should be aggregated first, instead of doing all of them?  For example first collect the allocation that has the most unaggregated fees?

aasseman replied:
Moved to #93

aasseman commented on tap_agent/src/tap/account.rs:
```diff
@@ -0,0 +1,846 @@
+        }
+    }
+
+    pub async fn start_last_rav_request(&self) {
+        *(self.inner.state.lock().await) = State::LastRavPending;
+        let mut rav_requester_task = self.rav_requester_task.lock().await;
+        if !Self::rav_requester_task_is_running(&rav_requester_task) {
+            *rav_requester_task = Some(tokio::spawn(Self::rav_requester(self.inner.clone())));
+        }
+    }
```
Yup, it's complex enough that we could have a risk of an unforeseen deadlock.

That was the best I could come up with to make sure that:
- Only one RAV request happens at a time
- The RAV request is async from all the rest
- I can set the "last rav" state asynchronously
- The RAV request can clear the state on its own (while living in its own async bubble)

Sorry, I guess I default back to Electrical Engineer mode in situations like this :grin:

aasseman commented on tap_agent/src/tap/account.rs:
```diff
@@ -0,0 +1,846 @@
+
+        let mut unaggregated_fees = inner.unaggregated_fees.lock().await;
+
+        // `COUNT(*)` will always return a value, so we don't need to check for `None`.
+        match res.count.unwrap() {
```
The actual reason I wanted the count was to run those checks:
https://github.com/graphprotocol/indexer-rs/blob/f7447f7f27a8436b74b4d20ce7abb878e576c354/tap_agent/src/tap/account.rs#L263-L270
That's indeed a bit overzealous, and a check like this should be more than enough:
```rs
ensure!(
    res.sum.is_none() ^ res.max.is_none(),
    "Exactly one of SUM(value) and MAX(id) is null. This should not happen."
);
```

aasseman replied:
Fixed in 8515c037b6d3db51de0812aee28a7ac3793efb09

aasseman commented on tap_agent/src/tap/account.rs:
```diff
@@ -0,0 +1,846 @@
+                .create_rav_request(inner.config.tap.rav_request_timestamp_buffer_ns)
+                .await?;
+
+            // TODO: Request compression and response decompression. Also a fancy user agent?
+            let client = HttpClientBuilder::default().build(&inner.sender_aggregator_endpoint)?;
```
We set some standards here: https://github.com/semiotic-ai/timeline-aggregation-protocol/blob/main/tap_aggregator/README.md

We recommend a max of 15,000 receipts / RAV request.

So here are the missing bits currently:
- In tap_core's Manager, we should be able to specify a maximim number of receipts / RAV request
- In TAP agent, we should trigger RAV requests not only a value, but also a number of receipts (< 15,000)

aasseman replied:
Created #96

aasseman commented on tap_agent/src/tap/account.rs:
```diff
@@ -0,0 +1,846 @@
+                    AND sender_address = $2
+            ) 
+            SELECT 
+                COUNT(*), 
```
I estimate that for a regular indexer, should be rarely above ~1000s, as there should be only the unaggregated receipts in there.

Also, I believe things could be optimized such that this query would happen only once at TAP Agent start.

aasseman commented:
Rebased onto latest main (at least so that the tests don't fail)

hopeyen commented on tap-agent/src/agent.rs:
```diff
@@ -0,0 +1,92 @@
+    let escrow_accounts = escrow_accounts(
+        escrow_subgraph,
+        config.ethereum.indexer_address,
+        Duration::from_secs(config.escrow_subgraph.escrow_syncing_interval),
```
duration should be from ms

aasseman replied:
Fixed in 8e6460f1b86ba67c1846fde6d5ae3345dbcc047c

hopeyen commented on tap-agent/src/tap/receipt_checks_adapter.rs:
```diff
@@ -0,0 +1,205 @@
+            transactions: Vec<AllocationResponse>,
+        }
+
+        timer(Duration::from_secs(escrow_subgraph_polling_interval)).map_with_retry(
```
Duration should be in `ms`?

aasseman replied:
Fixed in 8e6460f1b86ba67c1846fde6d5ae3345dbcc047c

hopeyen commented on tap-agent/src/tap/receipt_checks_adapter.rs:
```diff
@@ -0,0 +1,205 @@
+                    "Failed to check the escrow redeem status for allocation {} and sender {}: {}",
+                    allocation_id, sender_address, error
+                );
+                sleep(Duration::from_secs(escrow_subgraph_polling_interval).div_f32(2.))
```
this should be in `ms`?

aasseman replied:
Fixed in 8e6460f1b86ba67c1846fde6d5ae3345dbcc047c

hopeyen commented on tap-agent/src/tap/receipt_storage_adapter.rs:
```diff
@@ -0,0 +1,598 @@
+    type AdapterError = AdapterError;
+
+    /// We don't need this method in TAP Agent.
+    async fn store_receipt(&self, _receipt: ReceivedReceipt) -> Result<u64, Self::AdapterError> {
```
What component calls this function and `update_receipt_by_id`?

aasseman replied:
It the TapManager from tap_core: https://github.com/semiotic-ai/timeline-aggregation-protocol/blob/3d247c06380038d0a871a7427ce00e0406909160/tap_core/src/tap_manager/manager.rs#L80-L111.

But we don't need those since the storage of receipts is done only in `service`, and in a mush simplified way, so it doesn't even leverage the TapManager at all.

hopeyen reviewed (commented):
The overall structure makes sense to me. Would you be able to provide a list (maybe in docs or something) that tracks todo/unimplemented items so we can have a better view of the progress? I think we should also provide a link to the docs describing the lifetime of a receipt/rav

aasseman commented on tap-agent/src/config.rs:
```diff
@@ -0,0 +1,321 @@
+        env = "RAV_REQUEST_TIMESTAMP_BUFFER",
+        help = "Buffer (in ns) to add between the current time and the timestamp of the \
+        last unaggregated fee when triggering a RAV request.",
+        default_value_t = 1_000_000_000 // 1 second
```
Fixed in 8e6460f1b86ba67c1846fde6d5ae3345dbcc047c

aasseman commented on tap-agent/src/config.rs:
```diff
@@ -0,0 +1,321 @@
+        value_name = "rav-request-trigger-value",
+        env = "RAV_REQUEST_TRIGGER_VALUE",
+        help = "Value of unaggregated fees that triggers a RAV request (in GRT wei).",
+        default_value_t = 10_000_000_000_000_000_000 // 10 GRT
```
Fixed in 05cac7c316007bb6406169771fe44ab5b4b5e388

aasseman commented on tap-agent/src/tap/receipt_checks_adapter.rs:
```diff
@@ -0,0 +1,205 @@
+        Ok(true)
+    }
+
+    async fn is_valid_gateway_id(&self, gateway_id: Address) -> Result<bool, Self::AdapterError> {
```
Yup. Moving this to tap_core: https://github.com/semiotic-ai/timeline-aggregation-protocol/issues/188

aasseman commented on tap-agent/src/tap/escrow_adapter.rs:
```diff
@@ -0,0 +1,154 @@
+            .get(&sender)
+            .ok_or(AdapterError::AdapterError {
+                error: format!(
+                    "Gateway {} not found in escrow balances map, could not get available escrow.",
```
Fixed in 4f0ea80e4e9d8cdb3b4dfaec00b4b8fb719148ff

aasseman commented on tap_agent/src/tap/account.rs:
```diff
@@ -0,0 +1,846 @@
+            let updated_rows = sqlx::query!(
+                r#"
+                    UPDATE scalar_tap_latest_ravs
+                    SET is_last = true
```
Fixed in 1178c6386768c98b7efb8463c024c57286ef7d1b

aasseman commented:
> The overall structure makes sense to me. Would you be able to provide a list (maybe in docs or something) that tracks todo/unimplemented items so we can have a better view of the progress? I think we should also provide a link to the docs describing the lifetime of a receipt/rav

For the progress, I created a milestone here: https://github.com/graphprotocol/indexer-rs/milestone/1. I'm going to add all the todos there.

Jannis reviewed (approved):
This is good to go from my point of view. Thanks for addressing the feedback so quickly!

aasseman commented:
> This is good to go from my point of view. Thanks for addressing the feedback so quickly!

Thanks for taking the time to go through the big PR :pray: !
